feat: adds ephemeral auth guide (#151)
deepgram/deepgram-docs: fern/docs/token-based-authentication.mdx

---
title: Token-based Auth
subtitle: >-
  Generates an ephemeral temporary JSON Web Token (JWT) with a 30-second TTL (Time To Live) for Deepgram APIs.
slug: docs/token-based-authentication
---

> <Info>

Author: Will push up a change for this. Anytime we can drop Info call outs, I'm supportive of that.

If you need a short-lived ephemeral token, we highly recommend using [token-based authentication](/reference/ephemeral-auth-api/grant-token) over temporary API keys.
</Info>

> An ephemeral token is a secure, temporary authentication credential that provides time-limited access to Deepgram's APIs. These tokens have a 30 second Time To Live (TTL) and are designed specifically for short-lived, secure API access scenarios, particularly in client-side applications where long-term credential storage is not recommended. Unlike traditional API keys, ephemeral tokens are not persisted long-term and are ideal for real-time applications requiring secure, temporary access to Deepgram's services.

Reviewer: ```suggestion
An ephemeral token is a secure, temporary authentication credential that provides time-limited access to Deepgram's APIs. These tokens have a 30 second Time To Live (TTL) and are designed specifically for short-lived, secure API access scenarios, particularly in client-side applications where long-term credential storage is not recommended. Unlike traditional API keys, ephemeral tokens are ideal for real-time applications requiring secure, temporary access to Deepgram's services.
```

the persisted long-term thing is, ultimately, more of a benefit to us than to them. their benefit from the persistence change is that the ephemeral token api has better uptime targets from the company. hard to express that in documentation though, we don't currently make any _guarantees_ about uptime

<div style="width: 100%; display: flex; justify-content: center; margin: 2rem 0;">
<img src="../images/temp-auth-tokens.svg" alt="Token-based auth flow" width="600" height="446"/>
</div>

## Use Cases

> Ephemeral tokens are ideal for:
> - Real-time voice applications requiring secure client-side access
> - Temporary API access for third-party integrations
> - Secure sharing of API access with clients when combined with a backend proxy
> - WebSocket connections requiring authentication

Reviewer: ```suggestion
Ephemeral tokens are ideal for client-side/untrusted applications, because they allow clients to directly access the Deepgram API. Normal API keys, to be secure, require proxying traffic through your own servers where the API key is stored.

Direct access to the Deepgram API is particularly useful for client-side/untrusted applications that are more latency-sensitive, like realtime connections.
```

> ## Implementation Best Practices

Reviewer: idk if this section is needed. feels like this could be expressed through (commented) sample code. if i was going to say something about best practices (or rather, "things you could think about while you're here"), i'd say

> Creating ephemeral tokens still requires an API key (whose Voice API permissions those tokens will emulate). Because of this, you'll need a simple auth server to create and distribute tokens. This auth server is a perfect place to introduce any additional gating you need, like rate-limiting, or your own auth layer.

Author: @fakemonster  true, I did write this with the intention that we won't have a lot of sample code to begin with, and it helps our AI bot answer questions about this new functionality. I'd like to keep it if we can for those reasons.

Reviewer: If it's just a matter of having a bulleted list, that's fine, though it doesn't quite meet "best practices", it's more just "steps to implement" imo. i think best practices would get more into "what other functionality you could build onto this", "what are ways this won't work well for you" etc

also a minor detail, but "proxy" might be a little confusing for people, they may assume it's for proxying their voice api stuff. maybe better to say things like "create a backend service to generate tokens", "pass the token from your auth backend" etc?

Author: I'll make a few changes to this section to clarify.

    "expires_in": 30
}
  ```
</CodeGroup>

To use the JWT token, you can then pass it in the `Authorization` header of your request to any Deepgram API that supports token-based authentication using the `Bearer` scheme.

## Example

<CodeGroup>
 ```Curl cURL
  curl -X POST https://api.deepgram.com/v1/listen \
     -H "Authorization: Bearer YOUR_JWT_TOKEN"
  ```

</CodeGroup>

<Warning>
  Replace `YOUR_JWT_TOKEN` with the JWT token you received from the `/auth/grant` endpoint.
</Warning>

## SDK Implementation Examples

<CodeGroup>
  ```Typescript Typescript
  Coming Soon!
  ```
  ```Python Python
  Coming Soon!
  ```
  ```Go Go
  Coming Soon!
  ```
  ```.NET .NET
  Coming Soon!
  ```
</CodeGroup>

## FAQ

### Why do I get a Forbidden error?

`{"err_code":"FORBIDDEN","err_msg":"Insufficient permissions."}`

> The API key you use for the `/auth/grant` request needs to have at least Member permissions.

Reviewer: @jasonmajors is this true?

Reviewer 2: Yes this is true

To create a key with Member permissions. Login to the [Deepgram Console](https://console.deepgram.com/) and click on API Keys from the left sidebar. Click the "Create Key" button, then select "Advanced" options and choose "Member" from the permissions dropdown.

Alternatively, you can create a key with Member permissions using the [Create Key](/reference/management-api/keys/create) endpoint.

### Can a token have a longer TTL (Time To Live) than 30 seconds?

We currently don't support modifying the TTL value of ephemeral tokens.

### Can my websocket connection last longer than 30 seconds?

Yes! You only need the ephemeral token to be valid during the initial websocket connection to Deepgram's APIs. The websocket connection will then stay open as it would in any other case until you close it.

### Which Deepgram APIs can I use with an ephemeral tokens?

> These APIs will work with ephemeral tokens:

Reviewer: i kind of want to defer to @jasonmajors for terminology here, but I think it's worth making clear that API functionality (other than the agent weirdness) is based on the fact that ephemeral tokens are given restricted scopes

Author: Ok, let me know if anything needs to change. The intention was to clarify where temporary tokens can be used in terms of our APIs.

Reviewer 2: To match the API spec, 

Ephemeral tokens have usage::write permission for these apis:

Author: Update coming in now.

Reviewer: Even just a small note, like
> Tokens are only granted <whatever the scope is, jason knows> scopes, so they'll work with the following APIs:
> - <list of all the voice APIs except /agent>
>
> /agent API support coming soon!

helps imo

Reviewer 2: [The scope is `usage::write`](https://github.com/deepgram/console/pull/2390)

- [`/listen` REST API](/reference/speech-to-text-api/listen) - Speech to Text
- [`/listen` WebSocket API](/reference/speech-to-text-api/listen-streaming) - Speech to Text
- [`/speak` REST API](/reference/text-to-speech-api/speak) - Text to Speech
- [`/speak` WebSocket API](/reference/text-to-speech-api/speak-streaming) - Text to Speech
- [`/read` REST API](/reference/text-intelligence-api/text-read) - Text Intelligence

These APIs will not work with ephemeral tokens:

- The collection of [Management APIs](/reference/deepgram-api-overview)
- The [`/agent` API](/reference/voice-agent-api/agent) - Coming soon!

### Can I use this token for transcription batch requests?

> Yes! You can use this token with transcription batch requests. but if you're going to do this client-side there will be significant latency so this might not be optimal.

Reviewer 2: ```suggestion
Yes! You can use this token with batch requests. Just note that due to the short TTL you may need to request a new token before subsequent batch requests. Unless you have latency constraints, you're likely better off just making the batch request from inside the proxy 
```

### Are ephemeral tokens supported in the Deepgram SDKs?

Not yet, but SDK support is coming soon!

### Why should I use Ephemeral tokens over temporary API keys?

> Ephemeral tokens are preferred over temporary API keys because they are ephemeral: they are expired and removed after you are done using them.

Reviewer: ```suggestion
Ephemeral tokens don't require persistence, which confers a few benefits:
1. they are substantially quicker to create
2. they can be decentralized, and thus subject to stricter uptime than the management API can offer. For instance, a database maintenance period will make new API key creation unavailable, while existing API keys continue to work, alongside ephemeral token creation.
3. they aren't included in usage data, and so don't balloon your usage data exports like temporary API keys do
```

for #2, idk if this is necessarily the kind of stuff we like talking about out loud, but it's true, and arguably the most important reason to offer the feature. for #3, we don't quite have CSV exports yet, and people probably don't care too much about the JSON api (even if it's true for that too), but we're pretty close to CSV exports, in which case #3 becomes relevant

Reviewer 2: Here's how Natalie A just expressed it as we're drafting an email to customers. I like this personaly and I think it's accurate?
What this means for you:
1. Fewer Disruptions - Temporary tokens are more resilient to outages, reducing the chances of interruptions to your workflow — especially during planned maintenance or unexpected incidents.
2. Cleaner Console Experience-  Using temporary tokens means you’ll no longer see pages cluttered with expired API keys. Your usage summary and API keys views will be cleaner and easier to navigate.
3. Smoother Performance -  This update helps ensure your experience in the console remains fast and efficient, even as your usage scales.

Author: @czchapma  I can modify this section with this recommended language. Will push a change now for that.

> 

Reviewer 2: Suggestion for another FAQ
Can I track the usage from ephemeral tokens?
Yes! The ephemeral tokens generated have the same accessor as the API key used to generate them.

- [`/read` REST API](/reference/text-intelligence-api/text-read) - Text Intelligence

These APIs will not work with ephemeral tokens:

- The collection of [Management APIs](/reference/deepgram-api-overview)
- The [`/agent` API](/reference/voice-agent-api/agent) - Coming soon!

### Can I use this token for transcription batch requests?

Yes! You can use this token with batch requests. Just note that due to the short TTL you may need to request a new token before subsequent batch requests. Unless you have latency constraints, you're likely better off just making the batch request from inside the proxy

### Are ephemeral tokens supported in the Deepgram SDKs?

Not yet, but SDK support is coming soon!

### Why should I use Ephemeral tokens over temporary API keys?

- Fewer Disruptions - Temporary tokens are more resilient to outages, reducing the chances of interruptions to your workflow — especially during planned maintenance or unexpected incidents.
- Cleaner Deepgram Console Experience - Using temporary tokens means you’ll no longer see pages cluttered with expired API keys. Your usage summary and API keys views will be cleaner and easier to navigate.

> - Smoother Performance - This update helps ensure your experience in the console remains fast and efficient, even as your usage scales.

Reviewer: smoother performance is worth calling both in terms of console being faster _and_ "temp access" being faster. creating a token is absolutely faster than creating an API key, which means faster connection times for sockets etc

Author: Update coming in now

<div style="width: 100%; display: flex; justify-content: center; margin: 2rem 0;">
<img src="../images/temp-auth-tokens.svg" alt="Token-based auth flow" width="600" height="446"/>
</div>

## Use Cases

Ephemeral tokens are ideal for client-side/untrusted applications, because they allow clients to directly access the Deepgram API. Normal API keys, to be secure, require proxying traffic through your own servers where the API key is stored.

Direct access to the Deepgram API is particularly useful for client-side/untrusted applications that are more latency-sensitive, like realtime connections.

## Implementation Considerations

1. Create a backend service to generate ephemeral tokens
2. Generate tokens only when needed (e.g., at connection start)

> 3. Pass the token from your proxy backend to your client

Reviewer: ```suggestion
3. Pass the token from your backend to your client
```

  ```
  ```Python Python
  Coming Soon!
  ```
  ```Go Go
  Coming Soon!
  ```
  ```.NET .NET
  Coming Soon!
  ```
</CodeGroup>

## FAQ

### Why do I get a Forbidden error?

`{"err_code":"FORBIDDEN","err_msg":"Insufficient permissions."}`

The API key you use for the `/auth/grant` request needs to have at least Member permissions.
To create a key with Member permissions. Login to the [Deepgram Console](https://console.deepgram.com/) and click on API Keys from the left sidebar. Click the "Create Key" button, then select "Advanced" options and choose "Member" from the permissions dropdown.

Alternatively, you can create a key with Member permissions using the [Create Key](/reference/management-api/keys/create) endpoint.

### Can a token have a longer TTL (Time To Live) than 30 seconds?

We currently don't support modifying the TTL value of ephemeral tokens.

### Can my websocket connection last longer than 30 seconds?

Yes! You only need the ephemeral token to be valid during the initial websocket connection to Deepgram's APIs. The websocket connection will then stay open as it would in any other case until you close it.

### Which Deepgram APIs can I use with an ephemeral tokens?

> Ephemeral tokens have usage::write permission for these Deepgram APIs:

Reviewer: ```suggestion
Ephemeral tokens have usage:write permission for these Deepgram APIs:
```
